File: apps/web/lib/domain/room-sharing.ts
import { assignmentBlockDurations, type SnapshotAssignment } from "./contracts";

export interface RoomShareAssignmentGroup {
  key: string;
  assignments: SnapshotAssignment[];
}

export function roomShareAssignmentGroups(
  assignments: SnapshotAssignment[],
): RoomShareAssignmentGroup[] {
  const grouped = new Map<string, SnapshotAssignment[]>();
  for (const assignment of assignments) {
    const key = assignment.room_share_key?.trim();
    if (!key) continue;
    grouped.set(key, [...(grouped.get(key) ?? []), assignment]);
  }
  return [...grouped.entries()]
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([key, items]) => ({
      key,
      assignments: [...items].sort((left, right) =>
        left.id.localeCompare(right.id),
      ),
    }));
}

export function sharedRoomBlockDurations(
  assignments: SnapshotAssignment[],
): number[] {
  if (assignments.length < 2) return [];
  const durations = assignments.map(assignmentBlockDurations);
  const limit = Math.min(...durations.map((items) => items.length));
  const shared: number[] = [];
  for (let index = 0; index < limit; index += 1) {
    const duration = durations[0]?.[index];
    if (
      duration == null ||
      durations.some((items) => items[index] !== duration)
    ) {
      break;
    }
    shared.push(duration);
  }
  return shared;
}

export function roomShareBlockPairKey(
  leftBlockId: string,
  rightBlockId: string,
): string {
  return [leftBlockId, rightBlockId].sort().join("::");
}

export function sharedRoomBlockPairs(
  assignments: SnapshotAssignment[],
): Set<string> {
  const pairs = new Set<string>();
  for (const group of roomShareAssignmentGroups(assignments)) {
    if (group.assignments.length < 2) continue;
    const shared = sharedRoomBlockDurations(group.assignments);
    const leader = group.assignments[0]!;
    for (const follower of group.assignments.slice(1)) {
      shared.forEach((_duration, index) => {
        pairs.add(
          roomShareBlockPairKey(
            `${leader.id}:${index}`,
            `${follower.id}:${index}`,
          ),
        );
      });
    }
  }
  return pairs;
}

export function sharedRoomPeriodDiscount(
  assignments: SnapshotAssignment[],
): number {
  return roomShareAssignmentGroups(assignments).reduce((total, group) => {
    if (group.assignments.length < 2) return total;
    const sharedPeriods = sharedRoomBlockDurations(group.assignments).reduce(
      (sum, duration) => sum + duration,
      0,
    );
    return total + sharedPeriods * (group.assignments.length - 1);
  }, 0);
}

File: apps/solver/app/room_sharing.py
from collections import defaultdict

from app.models import Assignment


def room_share_assignment_groups(
    assignments: list[Assignment],
) -> list[tuple[str, list[Assignment]]]:
    grouped: dict[str, list[Assignment]] = defaultdict(list)
    for assignment in assignments:
        if assignment.room_share_key:
            grouped[assignment.room_share_key].append(assignment)
    return [
        (key, sorted(items, key=lambda item: item.id))
        for key, items in sorted(grouped.items())
    ]


def shared_room_block_durations(assignments: list[Assignment]) -> list[int]:
    if len(assignments) < 2:
        return []
    durations = [assignment.block_durations() for assignment in assignments]
    shared: list[int] = []
    for index in range(min(len(items) for items in durations)):
        duration = durations[0][index]
        if any(items[index] != duration for items in durations[1:]):
            break
        shared.append(duration)
    return shared


def room_share_block_pairs(assignments: list[Assignment]) -> list[tuple[str, str]]:
    pairs: list[tuple[str, str]] = []
    for _key, group in room_share_assignment_groups(assignments):
        if len(group) < 2:
            continue
        shared = shared_room_block_durations(group)
        leader = group[0]
        for follower in group[1:]:
            # A room-share pair means both lessons must run in the same room at the
            # same time. One teacher cannot teach both lessons simultaneously, so a
            # same-teacher pair is an invalid co-teaching constraint. Ignore that
            # derived room-share link and let the normal teacher/room constraints
            # place the lessons independently instead of making the whole model UNSAT.
            if leader.teacher_id == follower.teacher_id:
                continue
            for index in range(len(shared)):
                pairs.append((f"{leader.id}:{index}", f"{follower.id}:{index}"))
    return pairs


def room_share_block_pair_key(left_block_id: str, right_block_id: str) -> tuple[str, str]:
    return tuple(sorted((left_block_id, right_block_id)))
